### ISSUE TITLE
MIPS PE32 files won't decompile.

### ISSUE THREAD

**Issue by nihilus** [nihilus/snowman#38]
MIPS PE32 files won't decompile.

**Comment by techbliss** [comment]
It looks like the old issue https://github.com/nihilus/snowman/issues/31 happends also on windows.
tested on mips gdbserver from android ndk
https://mega.nz/#!Xsd2BBZC!z7vNvTHPFzSYSCdcCVrj7m-_ncra2GEoMIJXdAjbxzI

If you turn autodecompile off before opening file it dosent crash, but as soon you try decompile the bad string or code that tricker the error it crash.

It does show the error on this snapshot in the buttom of snowman.exe.
![lejjf](https://cloud.githubusercontent.com/assets/3592375/9976215/65037838-5edd-11e5-9ba1-d4a97a49c744.jpg)

Something about a indeks 2

**Comment by nihilus** [comment]
@techbliss isn't that an ELF file?

**Comment by techbliss** [comment]
yes this is a elf file, it crashes on all files when autodecompiling

**Comment by nihilus** [comment]
@techbliss I guess this should be a separate issue. Since this issue has to do of the PE-parsing for NT 4.0 binaries.

**Comment by techbliss** [comment]
Now i tried with the MIPS PE32 you supplied 
Same thing 
Here is the crash log
http://pastebin.com/kYqrgf1r

See the last calls 

```
00fccaa8 70bee1a9 0a687fbc 0a687fb8 70bedb53 msvcr100!abort+0x28
00fcd114 00a7eb01 00bf5d58 00bf5c88 00000064 msvcr100!_wassert+0x656
00fcd1bc 00a4f27b 68ff3fb0 0372017c 00cfcd20 snowman!nc::core::irgen::IRGenerator::generate+0x1c1
00fcd204 00a4d766 037d3e10 68ff3fd4 03720170 snowman!nc::core::MasterAnalyzer::createProgram+0xdb
00fcd260 0088ee74 0372017c 68ff3f2c 03720170 snowman!nc::core::MasterAnalyzer::decompile+0x86
00fcd298 00881349 0372017c 68ff3f70 03720170 snowman!nc::core::Driver::decompile+0x54
00fcd2c4 00881e1e 04ab76d0 00880861 68ff3f5c snowman!nc::gui::Decompilation::work+0x39
00fcd2cc 00880861 68ff3f5c 04ab76d0 00fcd318 snowman!nc::gui::Activity::run+0xe
00fcd2e8 0087b0e6 038158e8 68ff3e90 04b041c8 snowman!nc::gui::Command::delegate+0x71
00fcd324 00880bc1 68ff3ef8 036e7038 04ab76d0 snowman!nc::gui::DecompileAll::work+0x136
00fcd34c 00878915 68ff3e30 00000000 036e7018 snowman!nc::gui::Command::execute+0xa1
00fcd384 5c887769 036e7018 00000000 00000004 snowman!nc::gui::CommandQueue::executeNext+0xd5
WARNING: Stack unwind information not available. Following frames may be wrong.
00fcd3d8 5befda49 00000000 036e7018 02e6e540 QtCore4!QT::QMetaCallEvent::placeMetaCall+0x19
00fcd63c 5c878fcd 036e7018 04cd8ef8 5703d618 QtGui4!QT::QApplicationPrivate::notify_helper+0xb9
00fcd67c 5c87aaec 02c6ef78 04cd8ef8 5703d6a4 QtCore4!QT::QCoreApplication::notifyInternal+0x8d
00fcd6c0 5c89d7b5 00000000 00000000 02c6ef78 QtCore4!QT::QCoreApplicationPrivate::sendPostedEvents+0x1ec
00000000 00000000 00000000 00000000 00000000 QtCore4!QT::QEventDispatcherWin32::event+0x7b5
```

**Comment by nihilus** [comment]
Ah, some kind of assert() kicks in and kills the program.

**Comment by techbliss** [comment]
also little more detailed 

```
ChildEBP RetAddr  Args to Child              
00fcc0c4 76dc1286 00000003 00fcc674 00000001 ntdll!NtWaitForMultipleObjects+0xc
00fcc258 76dc1138 00000003 00fcc674 00000000 KERNELBASE!WaitForMultipleObjectsEx+0x136
00fcc274 7412a02a 00000003 00fcc674 00000000 KERNELBASE!WaitForMultipleObjects+0x18
00fcc6b0 74129ac6 00000000 00000001 00000000 kernel32!WerpReportFaultInternal+0x545
00fcc6c0 7410cf09 00fcc75c 76dd8e83 00fcc770 kernel32!WerpReportFault+0x7a
00fcc6c8 76dd8e83 00fcc770 00000001 5588ce4e kernel32!BasepReportFault+0x19
00fcc75c 70beaefd 00fcc770 00000000 00000000 KERNELBASE!UnhandledExceptionFilter+0x1b3
00fcca98 70bed6fd 00000003 40000015 00000001 msvcr100!_call_reportfault+0x107
00fccaa8 70bee1a9 0a687fbc 0a687fb8 70bedb53 msvcr100!abort+0x28
00fcd114 00a7eb01 00bf5d58 00bf5c88 00000064 msvcr100!_wassert+0x656
00fcd1bc 00a4f27b 68ff3fb0 0372017c 00cfcd20 snowman!nc::core::irgen::IRGenerator::generate+0x1c1 [d:\snowman-master\src\nc\core\irgen\irgenerator.cpp @ 100]
00fcd204 00a4d766 037d3e10 68ff3fd4 03720170 snowman!nc::core::MasterAnalyzer::createProgram+0xdb [d:\snowman-master\src\nc\core\masteranalyzer.cpp @ 70]
00fcd260 0088ee74 0372017c 68ff3f2c 03720170 snowman!nc::core::MasterAnalyzer::decompile+0x86 [d:\snowman-master\src\nc\core\masteranalyzer.cpp @ 229]
00fcd298 00881349 0372017c 68ff3f70 03720170 snowman!nc::core::Driver::decompile+0x54 [d:\snowman-master\src\nc\core\driver.cpp @ 101]
00fcd2c4 00881e1e 04ab76d0 00880861 68ff3f5c snowman!nc::gui::Decompilation::work+0x39 [d:\snowman-master\src\nc\gui\decompilation.cpp @ 45]
00fcd2cc 00880861 68ff3f5c 04ab76d0 00fcd318 snowman!nc::gui::Activity::run+0xe [d:\snowman-master\src\nc\gui\activity.cpp @ 32]
00fcd2e8 0087b0e6 038158e8 68ff3e90 04b041c8 snowman!nc::gui::Command::delegate+0x71 [d:\snowman-master\src\nc\gui\command.cpp @ 71]
00fcd324 00880bc1 68ff3ef8 036e7038 04ab76d0 snowman!nc::gui::DecompileAll::work+0x136 [d:\snowman-master\src\nc\gui\decompileall.cpp @ 57]
00fcd34c 00878915 68ff3e30 00000000 036e7018 snowman!nc::gui::Command::execute+0xa1 [d:\snowman-master\src\nc\gui\command.cpp @ 55]
00fcd384 5c887769 036e7018 00000000 00000004 snowman!nc::gui::CommandQueue::executeNext+0xd5 [d:\snowman-master\src\nc\gui\commandqueue.cpp @ 85]
WARNING: Stack unwind information not available. Following frames may be wrong.
00fcd3d8 5befda49 00000000 036e7018 02e6e540 QtCore4!QT::QMetaCallEvent::placeMetaCall+0x19
00fcd63c 5c878fcd 036e7018 04cd8ef8 5703d618 QtGui4!QT::QApplicationPrivate::notify_helper+0xb9
00fcd67c 5c87aaec 02c6ef78 04cd8ef8 5703d6a4 QtCore4!QT::QCoreApplication::notifyInternal+0x8d
00fcd6c0 5c89d7b5 00000000 00000000 02c6ef78 QtCore4!QT::QCoreApplicationPrivate::sendPostedEvents+0x1ec
00000000 00000000 00000000 00000000 00000000 QtCore4!QT::QEventDispatcherWin32::event+0x7b5
```

```
00fcd1bc 00a4f27b 68ff3fb0 0372017c 00cfcd20 snowman!nc::core::irgen::IRGenerator::generate+0x1c1 [d:\snowman-master\src\nc\core\irgen\irgenerator.cpp @ 100]
```

**Comment by nihilus** [comment]
I hate that assert on line 100... in irgenerator.cpp
Try to charge the #if 1 to #if 0

**Comment by techbliss** [comment]
new build with 

```
#if 1 > #if 0
```

now it decompile automatic without error on the gdb server MIPS
MIPS PE32 does crash but this crash happends later.
Let me anaylze this crash

**Comment by techbliss** [comment]
okay here is the dump analyze of MIPS PE32 one 

```
0:000> .ecxr
eax=009ec220 ebx=00000003 ecx=00000000 edx=00000000 esi=70c163b8 edi=00000000
eip=70bed6fd esp=009ec4f4 ebp=009ecb6c iopl=0         nv up ei pl nz ac pe nc
cs=0023  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00200216
msvcr100!abort+0x28:
70bed6fd 83c40c          add     esp,0Ch
0:000> kb
  *** Stack trace for last set context - .thread/.cxr resets it
ChildEBP RetAddr  Args to Child              
009ec500 70bee1a9 0b9a2128 0b9a2128 00000000 msvcr100!abort+0x28
009ecb6c 014d3f77 0161a5d8 016199c8 000001fc msvcr100!_wassert+0x656
009ecbcc 014d48fc 009ecc64 00000000 0fd9dce8 snowman!nc::core::ir::cgen::DefinitionGenerator::makeExpression+0x167 [d:\hilde\src\nc\core\ir\cgen\definitiongenerator.cpp @ 508]
009ecd8c 014d5759 0d975dd8 0fd9dce8 00000000 snowman!nc::core::ir::cgen::DefinitionGenerator::makeStatements+0x23c [d:\hilde\src\nc\core\ir\cgen\definitiongenerator.cpp @ 271]
009ecdb8 014d47c1 00000008 0fd9dce8 00000000 snowman!nc::core::ir::cgen::DefinitionGenerator::makeStatements+0x79 [d:\hilde\src\nc\core\ir\cgen\definitiongenerator.cpp @ 488]
009ecf78 014d5b27 0d9761f8 0fd9dce8 00000000 snowman!nc::core::ir::cgen::DefinitionGenerator::makeStatements+0x101 [d:\hilde\src\nc\core\ir\cgen\definitiongenerator.cpp @ 250]
009ed014 014b5a3b 009ed028 1779afd3 009ed16c snowman!nc::core::ir::cgen::DefinitionGenerator::createDefinition+0x387 [d:\hilde\src\nc\core\ir\cgen\definitiongenerator.cpp @ 157]
009ed0e8 014b5bb8 0aecea48 1779ae7f 03679b48 snowman!nc::core::ir::cgen::CodeGenerator::makeFunctionDefinition+0x5b [d:\hilde\src\nc\core\ir\cgen\codegenerator.cpp @ 243]
009ed144 0146f2cd 1779ad3f 035ebf0c 0171f048 snowman!nc::core::ir::cgen::CodeGenerator::makeCompilationUnit+0xd8 [d:\hilde\src\nc\core\ir\cgen\codegenerator.cpp @ 64]
009ed204 0146ade6 035ebf0c 1779ad5b 035ebf00 snowman!nc::core::MasterAnalyzer::generateTree+0xfd [d:\hilde\src\nc\core\masteranalyzer.cpp @ 217]
009ed260 012aecd4 035ebf0c 1779ada3 035ebf00 snowman!nc::core::MasterAnalyzer::decompile+0x1d6 [d:\hilde\src\nc\core\masteranalyzer.cpp @ 265]
009ed298 012a11a9 035ebf0c 1779adff 035ebf00 snowman!nc::core::Driver::decompile+0x54 [d:\hilde\src\nc\core\driver.cpp @ 101]
009ed2c4 012a1c7e 03692730 012a06c1 1779add3 snowman!nc::gui::Decompilation::work+0x39 [d:\hilde\src\nc\gui\decompilation.cpp @ 45]
009ed2cc 012a06c1 1779add3 03692730 009ed318 snowman!nc::gui::Activity::run+0xe [d:\hilde\src\nc\gui\activity.cpp @ 32]
009ed2e8 0129af46 07611b78 1779ac1f 07609a58 snowman!nc::gui::Command::delegate+0x71 [d:\hilde\src\nc\gui\command.cpp @ 71]
009ed324 012a0a21 1779ac77 035b9e00 03692730 snowman!nc::gui::DecompileAll::work+0x136 [d:\hilde\src\nc\gui\decompileall.cpp @ 57]
009ed34c 01298775 1779acbf 00000000 035b9de0 snowman!nc::gui::Command::execute+0xa1 [d:\hilde\src\nc\gui\command.cpp @ 55]
009ed384 66837769 035b9de0 00000000 00000004 snowman!nc::gui::CommandQueue::executeNext+0xd5 [d:\hilde\src\nc\gui\commandqueue.cpp @ 85]
WARNING: Stack unwind information not available. Following frames may be wrong.
009ed3d8 6454da49 00000000 035b9de0 01113070 QtCore4!QT::QMetaCallEvent::placeMetaCall+0x19
009ed614 66828fcd 035b9de0 077fbd28 37005038 QtGui4!QT::QApplicationPrivate::notify_helper+0xb9
009ed654 6682aaec 00c18fb0 077fbd28 370050f4 QtCore4!QT::QCoreApplication::notifyInternal+0x8d
009ed698 6684d7b5 00000000 00000000 00c18fb0 QtCore4!QT::QCoreApplicationPrivate::sendPostedEvents+0x1ec
00000000 00000000 00000000 00000000 00000000 QtCore4!QT::QEventDispatcherWin32::event+0x7b5
```

```
009ecbcc 014d48fc 009ecc64 00000000 0fd9dce8 snowman!nc::core::ir::cgen::DefinitionGenerator::makeExpression+0x167 [d:\hilde\src\nc\core\ir\cgen\definitiongenerator.cpp @ 508]
```

```
if (const Jump *jump = statement->asJump()) {
                assert(jump == basicNode->basicBlock()->getJump());

                expression = makeExpression(jump->condition());

                assert((jump->thenTarget().basicBlock() == thenBB && jump->elseTarget().basicBlock() == elseBB) ||
                       (jump->thenTarget().basicBlock() == elseBB && jump->elseTarget().basicBlock() == thenBB));
```

**Comment by nihilus** [comment]
Fixed in ab247f039249b28554ad6c89a56c21ccdaef51f8
